larq_compute_engine/core/indirect_bgemm/kernel_4x2_portable.h

```diff
@@ -0,0 +1,243 @@
+    std::int32_t acc_20 = 0, acc_21 = 0;
+    std::int32_t acc_30 = 0, acc_31 = 0;
+
+    std::int32_t k_size_index = conv_kernel_size;
+    do {
+      const TBitpacked* activations_ptr_0 = indirection_buffer[0];
+      const TBitpacked* activations_ptr_1 = indirection_buffer[1];
+      indirection_buffer += 2;
+
+      std::int32_t c_in_index = channels_in;
+      do {
+        const TBitpacked w_0 = weights_ptr[0];
+        const TBitpacked w_1 = weights_ptr[1];
+        const TBitpacked w_2 = weights_ptr[2];
+        const TBitpacked w_3 = weights_ptr[3];
+        weights_ptr += 4;
+
+        const TBitpacked a_0 = *activations_ptr_0++;
+        const TBitpacked a_1 = *activations_ptr_1++;
+
+        acc_00 += xor_popcount(w_0, a_0);
+        acc_10 += xor_popcount(w_1, a_0);
+        acc_20 += xor_popcount(w_2, a_0);
+        acc_30 += xor_popcount(w_3, a_0);
+        acc_01 += xor_popcount(w_0, a_1);
+        acc_11 += xor_popcount(w_1, a_1);
+        acc_21 += xor_popcount(w_2, a_1);
+        acc_31 += xor_popcount(w_3, a_1);
+      } while (--c_in_index > 0);
+    } while (--k_size_index > 0);
+
+    output_col_0 |= TBitpacked(output_transform.Run(acc_00, c_out_index))
+                    << (c_out_index % bitpacking_bitwidth);
+    output_col_0 |= TBitpacked(output_transform.Run(acc_10, c_out_index + 1))
+                    << ((c_out_index + 1) % bitpacking_bitwidth);
+    output_col_0 |= TBitpacked(output_transform.Run(acc_20, c_out_index + 2))
+                    << ((c_out_index + 2) % bitpacking_bitwidth);
+    output_col_0 |= TBitpacked(output_transform.Run(acc_30, c_out_index + 3))
+                    << ((c_out_index + 3) % bitpacking_bitwidth);
+    output_col_1 |= TBitpacked(output_transform.Run(acc_01, c_out_index))
+                    << (c_out_index % bitpacking_bitwidth);
+    output_col_1 |= TBitpacked(output_transform.Run(acc_11, c_out_index + 1))
+                    << ((c_out_index + 1) % bitpacking_bitwidth);
+    output_col_1 |= TBitpacked(output_transform.Run(acc_21, c_out_index + 2))
+                    << ((c_out_index + 2) % bitpacking_bitwidth);
+    output_col_1 |= TBitpacked(output_transform.Run(acc_31, c_out_index + 3))
+                    << ((c_out_index + 3) % bitpacking_bitwidth);
+
+    indirection_buffer -= 2 * conv_kernel_size;
+    c_out_index += 4;
+
+    // Write the bitpacked columns whenever they are full, or if we've computed
+    // the last output column value.
+    if (c_out_index % bitpacking_bitwidth == 0 || c_out_index >= channels_out) {
+      // If this is a 'leftover output channel' block (because the number of
+      // output channels isn't a multiple of four) then zero-out the extra bits.
+      if (c_out_index % bitpacking_bitwidth != 0) {
+        output_col_0 &=
+            (TBitpacked(1) << (channels_out % bitpacking_bitwidth)) - 1;
+        output_col_1 &=
+            (TBitpacked(1) << (channels_out % bitpacking_bitwidth)) - 1;
+      }
```

Comment:
The extra bits should already be zero right? Because you set the entire word to zero at the start and at the end of every iteration (otherwise the `|= (1 << i);` wouldn't work).

Reply:
I don't think so necessarily, because the kernel computes four bits at once, and there might be only two output channels.

Reply:
Ah yes, makes sense :+1: